### PR TITLE
Jcs/session

### PR DESCRIPTION
**Pull request by jSylvestre**
Cache calls to team permissions for the security service

### CODE REVIEW

File: Keas.Mvc/Services/SessionManagementService.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+using System.Threading.Tasks;
+using Keas.Core.Data;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Rewrite.Internal.ApacheModRewrite;
+using Microsoft.EntityFrameworkCore;
+using Newtonsoft.Json;
+
+namespace Keas.Mvc.Services
+{
+    public interface ISessionManagementService
```
Comment (jpknoll) [inline Keas.Mvc/Services/SessionManagementService.cs]: pick a more specific name

RolesSessionsManager

File: Keas.Mvc/Services/SessionManagementService.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        /// Will return empty array if no roles.
+        /// </summary>
+        /// <param name="slug"></param>
+        /// <returns></returns>
+        public async Task<string[]> GetTeamRoleNames(string slug)
+        {
+            var userId = _contextAccessor.HttpContext.User.Identity.Name;
+            if (_contextAccessor.HttpContext.Session.GetString($"UserId:{userId}") != userId)
+            {
+                _contextAccessor.HttpContext.Session.Clear(); //Abandon?
```
Comment (jpknoll) [inline Keas.Mvc/Services/SessionManagementService.cs]: clears the entire session, logout?

**Comment by jSylvestre** [comment]
image

File: Keas.Mvc/Services/RolesSessionsManager.cs [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+using System.Threading.Tasks;
+using Keas.Core.Data;
+using Microsoft.AspNetCore.Http;
+using Microsoft.EntityFrameworkCore;
+using Newtonsoft.Json;
+using Serilog;
+
+namespace Keas.Mvc.Services
+{
+    public interface IRolesSessionsManager
```
Comment (srkirkland) [inline Keas.Mvc/Services/RolesSessionsManager.cs]: Something drastic w/ naming here -- how about just `IRolesManager`?  The interface shouldn't care if the implementation is actually done via session or cache or cookies or w/e.  Then we'd want to change the `ClearSessionRoles` method to just `ClearRoles` so it's more generic.

File: Keas.Mvc/Services/RolesSessionsManager.cs [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    {
+        Task<string[]> GetTeamRoleNames(string slug);
+        Task<string[]> GetTeamOrAdminRoleNames(string slug);
+        Task<string[]> GetSystemRoleNames();
+        void ClearSessionRoles();
+
+    }
+
+    public class RolesSessionsManager : IRolesSessionsManager
```
Comment (srkirkland) [inline Keas.Mvc/Services/RolesSessionsManager.cs]: For the implementation, i don't like the double plural.  There aren't multiple sessions so maybe just `RolesSessionManager`?

File: Keas.Mvc/Services/RolesSessionsManager.cs [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+            var userId = _contextAccessor.HttpContext.User.Identity.Name;
+            var systemRole = roleContainer.SystemRoles;
+            if (systemRole == null)
+            {
+                systemRole = new Team
+                {
+                    TeamName = "System",
+                    TeamRoles = await _dbContext.SystemPermissions
+                        .AsNoTracking()
+                        .Where(a => a.UserId == userId).Select(a => a.Role.Name).Distinct().ToArrayAsync(),
```
Comment (srkirkland) [inline Keas.Mvc/Services/RolesSessionsManager.cs]: @jSylvestre It's safe to leave but I'm not sure you need asNoTracking() if you are going to .Select() project out.  There is nothing to track anyway

File: Keas.Mvc/Services/RolesSessionsManager.cs (hunk 3 above)
Comment (jpknoll) [reply Keas.Mvc/Services/RolesSessionsManager.cs]: Yeah, this is the right direction.

File: Keas.Mvc/Services/RolesSessionsManager.cs (hunk 4 above)
Comment (jpknoll) [reply Keas.Mvc/Services/RolesSessionsManager.cs]: My concern around RolesManager is that name already exists for the EF Roles Manager.

File: Keas.Mvc/Services/RolesSessionsManager.cs (hunk 5 above)
Comment (jpknoll) [reply Keas.Mvc/Services/RolesSessionsManager.cs]: Yeah, if the query is returning just a list of strings, he can probably leave it out. 

I'd even consider raw SQL for this part of the application.

**Comment by jSylvestre** [comment]
close issue #618

File: Keas.Mvc/Services/TeamsManager.cs [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+namespace Keas.Mvc.Services
+{
+    public interface ITeamsManager
+    {
+        Task<string[]> GetTeamRoleNames(string slug);
+        Task<string[]> GetTeamOrAdminRoleNames(string slug);
+        Task<string[]> GetSystemRoleNames();
+        Task<Dictionary<string, string>> GetMyTeams();
+        void ClearTeamSession();
```
Comment (srkirkland) [inline Keas.Mvc/Services/TeamsManager.cs]: still don't like this having the session as part of the name, but it's not that bad

File: Keas.Mvc/Services/TeamsManager.cs [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+            public List<Team> Teams { get; set; }
+
+            public Team SystemRoles { get; set; }
+            public Dictionary<string, string> TeamDictionary { get; set; }
+        }
+
+        public class Team
+        {
+            public string TeamName { get; set; }
+            public string[] TeamRoles { get; set; }
```
Comment (srkirkland) [inline Keas.Mvc/Services/TeamsManager.cs]: Generally I don't think it's necessary to add the classname to the properties of that class, like I'd rather you have `roles.team.name` than `roles.team.teamName`.

**Pull request merged**